examples/step-55/step-55.cc

```diff
@@ -0,0 +1,1212 @@
+      SolverCG<>    cg (solver_control);
+
+
+      // This takes care of the mass matrix
+      dst.block(1) = 0.0;
+      cg.solve(pressure_mass_matrix,
+               dst.block(1), src.block(1),
+               mp_preconditioner);
+      n_iterations_S_ += solver_control.last_step();
+
+
+      dst.block(1) *= -1.0;
+    }
+
+    // Apply the top right block
+    {
+      stokes_matrix.block(0,1).vmult(utmp, dst.block(1)); //B^T
+      utmp*=-1.0;
+      utmp.add(src.block(0));
+    }
+
+    // Now either solve with the top left block (if do_solve_A==true)
+    // or just apply one preconditioner sweep (for the first few
+    // iterations of our two-stage outer GMRES iteration)
+    if (do_solve_A == true)
+      {
+        SolverControl solver_control(10000, utmp.l2_norm()*1e-2);
+        SolverCG<>    cg (solver_control);
+
+        dst.block(0) = 0.0;
+        cg.solve(stokes_matrix.block(0,0), dst.block(0), utmp,
+                 a_preconditioner);
+        n_iterations_A_ += solver_control.last_step();
+      }
+    else
+      {
+        a_preconditioner.vmult (dst.block(0), utmp);
+        n_iterations_A_ += 1;
+      }
+  }
+
+  // This structure is similar to Step-22 except the 2D and 3D case both use ILU
+  template <int dim>
+  struct InnerPreconditioner;
+
+  template <>
+  struct InnerPreconditioner<2>
+  {
+    typedef SparseILU<double> type;
+  };
+
+  template <>
+  struct InnerPreconditioner<3>
+  {
+    typedef SparseILU<double> type;
+  };
+
+  template <int dim>
+  class StokesProblem
+  {
+  public:
+    StokesProblem (const unsigned int degree);
+    void run ();
+
+  private:
+    void setup_dofs ();
+    void assemble_system ();
+    void assemble_multigrid ();
+    void solve ();
+    void output_results (const unsigned int refinement_cycle) const;
+    void refine_mesh ();
+
+    const unsigned int   degree;
+
+    Triangulation<dim>   triangulation;
+    FESystem<dim>        fe;
+    FESystem<dim>        velocity_fe;
+    DoFHandler<dim>      dof_handler;
+    DoFHandler<dim>      velocity_dof_handler;
+
+    ConstraintMatrix     constraints;
+    ConstraintMatrix     velocity_constraints;
+
+    ConstraintMatrix     hanging_node_constraints;
+    ConstraintMatrix     velocity_hanging_node_constraints;
+
+    BlockSparsityPattern      sparsity_pattern;
+    BlockSparsityPattern      sparsity_pattern_velocity;
+    BlockSparseMatrix<double> system_matrix;
+
+    BlockVector<double> solution;
+    BlockVector<double> system_rhs;
+
+    std::vector<double> dof_indices;
+    std::vector<double> vel_dof_indices;
+
+    MGLevelObject<SparsityPattern>        mg_sparsity_patterns;
+    MGLevelObject<SparseMatrix<double> > mg_matrices;
+    MGLevelObject<SparseMatrix<double> > mg_interface_matrices;
+    MGConstrainedDoFs                     mg_constrained_dofs;
+
+    TimerOutput computing_timer;
+  };
+
+  template <int dim>
+  class BoundaryValues : public Function<dim>
```

Comment:
Hmm.. none of our usual examples will be as good since we don't have our usual equations.  What about the kind of example Leo has all of his students doing with the square cylinder in the box or whatever it is?  Would that be of any interest here?